### ISSUE TITLE
:terminal emacs doesn't work

### ISSUE THREAD

**Issue by mattn** [vim/vim#1913]
```
:terminal emacs
```
got error
```
emacs: Could not open file: /dev/tty
```
but spawn emacs from `:terminal bash` works fine.

**Comment by mattn** [comment]
vim doesn't bind slave fd for stdin/stdout/stderr.

![terminal](https://user-images.githubusercontent.com/10111/28866321-8d65a5f4-77ad-11e7-8c7f-8b004a7a1c6c.gif)

```diff
diff --git a/src/os_unix.c b/src/os_unix.c
index fb2628c..e96c3ff 100644
--- a/src/os_unix.c
+++ b/src/os_unix.c
@@ -5361,10 +5361,7 @@ mch_job_start(char **argv, job_T *job, jobopt_T *options)
 	if (fd_err[1] >= 0)
 	    close(fd_err[1]);
 	if (pty_master_fd >= 0)
-	{
 	    close(pty_master_fd); /* not used */
-	    close(pty_slave_fd); /* duped above */
-	}
 
 	if (null_fd >= 0)
 	    close(null_fd);
@@ -5378,6 +5375,10 @@ mch_job_start(char **argv, job_T *job, jobopt_T *options)
 	     * unless run by root) */
 	    ioctl(pty_slave_fd, TIOCSCTTY, (char *)NULL);
 #  endif
+	    while (dup2(pty_slave_fd, STDIN_FILENO) == -1 && errno == EBUSY);
+	    while (dup2(pty_slave_fd, STDOUT_FILENO) == -1 && errno == EBUSY);
+	    while (dup2(pty_slave_fd, STDERR_FILENO) == -1 && errno == EBUSY);
+	    if (pty_slave_fd > 2) close(pty_slave_fd);
 	}
 
 	/* See above for type of argv. */
```

**Comment by brammool** [comment]
Yasuhiro wrote:

> vim doesn't bind slave fd for stdin/stdout/stderr.
> 
> ![terminal](https://user-images.githubusercontent.com/10111/28866321-8d65a5f4-77ad-11e7-8c7f-8b004a7a1c6c.gif)

This deserves an explanation.

Why use dup2() after already doing a normal dup()?

Why would dup2() be needed, this is in the child were we don't really
expect any race conditions.

Actually, the problem is that pty_slave_fd is closed before doing an
ioctl() on it.  Even though it's duped does not mean that works.

It's probably best to do the ioctl() before dup'ing.


-- 
ARTHUR:  Well, I AM king...
DENNIS:  Oh king, eh, very nice.  An' how'd you get that, eh?  By exploitin'
         the workers -- by 'angin' on to outdated imperialist dogma which
         perpetuates the economic an' social differences in our society!  If
         there's ever going to be any progress--
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
closed by [v8.0.0842](https://github.com/vim/vim/releases/tag/v8.0.0842)
